js/src/oai.ts

```diff
@@ -1,24 +1,154 @@
-// TODO REPLACE WITH autoevals version
-// https://github.com/braintrustdata/braintrust/issues/218
+import { currentSpan, getCurrentUnixTimestamp, Span } from "./logger";
 
-import { Configuration, OpenAIApi } from "openai";
+interface ChatLike {
+  completions: any;
+}
+interface OpenAILike {
+  chat: ChatLike;
+}
 
-let _openai: OpenAIApi | null = null;
-export function openAI() {
-  if (_openai === null && process.env.OPENAI_API_KEY) {
-    const config = new Configuration({ apiKey: process.env.OPENAI_API_KEY });
-    _openai = new OpenAIApi(config);
+/**
+ * Wrap an `OpenAI` object (created with `new OpenAI(...)`) to add tracing. If Braintrust is
+ * not configured, this is a no-op
+ *
+ * Currently, this only supports the `v4` API.
+ *
+ * @param openai
+ * @returns The wrapped `OpenAI` object.
+ */
+export function wrapOpenAI<T extends object>(openai: T): T {
+  if ((openai as any)?.chat?.completions?.create) {
+    return wrapOpenAIv4(openai as any) as T;
+  } else {
+    console.warn("Unsupported OpenAI library (potentially v3). Not wrapping.");
+    return openai;
   }
-  return _openai;
 }
 
-export async function chatCompletion(args: any) {
-  const openai = openAI();
-  if (openai === null) {
-    throw new Error("OPENAI_API_KEY not set");
+export function wrapOpenAIv4<T extends OpenAILike>(openai: T): T {
+  let completionProxy = new Proxy(openai.chat.completions, {
+    get(target, name, receiver) {
+      const baseVal = Reflect.get(target, name, receiver);
+      if (name === "create") {
+        return wrapChatCompletionNonStreaming(baseVal.bind(target));
+      }
+      return baseVal;
+    },
+  });
+  let chatProxy = new Proxy(openai.chat, {
+    get(target, name, receiver) {
+      if (name === "completions") {
+        return completionProxy;
+      }
+      return Reflect.get(target, name, receiver);
+    },
+  });
+  let proxy = new Proxy(openai, {
+    get(target, name, receiver) {
+      if (name === "chat") {
+        return chatProxy;
+      }
+      return Reflect.get(target, name, receiver);
+    },
+  });
+
+  return proxy;
+}
+
+type ChatParams = {
+  messages: unknown;
+  stream?: boolean | null;
+};
+
+interface NonStreamingChatResponse {
+  choices: any[];
+  usage:
+    | {
+        total_tokens: number;
+        prompt_tokens: number;
+        completion_tokens: number;
+      }
+    | undefined;
+}
+
+// TODO: Mock this up better
+type StreamingChatResponse = any;
+
+function wrapChatCompletionNonStreaming<
+  P extends ChatParams,
+  C extends NonStreamingChatResponse | StreamingChatResponse
+>(completion: (params: P) => Promise<C>): (params: P) => Promise<any> {
+  return async (params: P) => {
+    if (params.stream) {
+      const { messages, ...rest } = params;
+      const span = currentSpan().startSpan("OpenAI Chat Completion", {
```

Comment:
should we use the global method `startSpan`? `currentSpan` will not work if your active object is an experiment/logger. Also could we share the event content across both branches?
```
const { messages, ...rest } = params;
const initEvent = {
  input: messages,
  metadata: { ...rest }
};
...
```

Reply:
Yeah good call